File: 3.x.x/crates-utopia/emails/src/validator/email.rs
use serde_json::Value;
use utopia_validators::{Validator, ValueType};

use crate::Email as EmailParser;

/// Validate that a value is a valid email address (PHP `Utopia\Emails\Validator\Email`).
#[derive(Debug, Clone)]
pub struct Email {
    allow_empty: bool,
}

impl Email {
    /// PHP `new Email($allowEmpty = false)`.
    pub fn new(allow_empty: bool) -> Self {
        Self { allow_empty }
    }
}

impl Default for Email {
    fn default() -> Self {
        Self::new(false)
    }
}

impl Validator for Email {
    fn description(&self) -> String {
        "Value must be a valid email address".into()
    }

    fn is_array(&self) -> bool {
        false
    }

    fn value_type(&self) -> ValueType {
        ValueType::String
    }

    fn is_valid(&self, value: &Value) -> bool {
        let Some(raw) = value.as_str() else {
            return false;
        };
        if self.allow_empty && raw.is_empty() {
            return true;
        }
        EmailParser::new(raw)
            .map(|email| email.is_valid())
            .unwrap_or(false)
    }
}

File: packages/emails/src/Validator/Email.php
<?php

namespace Utopia\Emails\Validator;

use Utopia\Emails\Email as EmailParser;
use Utopia\Validator;

/**
 * Email
 *
 * Validate that a value is a valid email address
 */
class Email extends Validator
{
    protected bool $allowEmpty;

    public function __construct(bool $allowEmpty = false)
    {
        $this->allowEmpty = $allowEmpty;
    }

    /**
     * Get Description
     *
     * Returns validator description
     */
    public function getDescription(): string
    {
        return 'Value must be a valid email address';
    }

    /**
     * Is valid
     *
     * Validation will pass when $value is a valid email address
     *
     * @param  mixed  $value
     */
    public function isValid($value): bool
    {
        if (! is_string($value)) {
            return false;
        }

        if ($this->allowEmpty && \strlen($value) === 0) {
            return true;
        }

        try {
            $email = new EmailParser($value);

            return $email->isValid();
        } catch (\Exception $e) {
            return false;
        }
    }

    /**
     * Is array
     *
     * Function will return true if object is array.
     */
    public function isArray(): bool
    {
        return false;
    }

    /**
     * Get Type
     *
     * Returns validator type.
     */
    public function getType(): string
    {
        return self::TYPE_STRING;
    }
}
